cosmos/lib/pagination.go

```diff
@@ -0,0 +1,37 @@
+package lib
+
+import (
+	"github.com/cosmos/cosmos-sdk/types/query"
+)
+
+type PageResponse interface {
+	GetPagination() *query.PageResponse
+}
+
+func getNextPaginationKey(response PageResponse) []byte {
+	if response == nil {
+		return nil
+	}
+	pagination := response.GetPagination()
+	if pagination == nil {
+		return nil
+	}
+	return pagination.NextKey
+}
+
+// Paginate returns the aggregate responses of a paginated cosmos query
+func Paginate[T PageResponse](
+	responseFetcher func(paginationKey []byte) (T, error),
+) (responses []T, err error) {
+	var response T
+
+	for paginationKey, i := []byte{}, 0; len(paginationKey) > 0 || i == 0; paginationKey, i = getNextPaginationKey(response), i+1 {
```

Comment:
I'm not sure about the unbounded loop here (only termination condition being `paginationKey`, regardless of number of iterations)